Modules/Keychain/Sources/Secured.swift

```
@@ -0,0 +1,173 @@
+            kSecAttrService as String: key
+        ]
+    }
+
+    private let key: String
+    private let keychain: KeychainProtocol
+
+    /// The default initializer for `Secured`
+    ///
+    /// - Parameter key: The key associated with storing the value inside the Keychain
+    public init(key: String, keychain: KeychainProtocol = Keychain.default) {
+        self.key = key
+        self.keychain = keychain
+
+        do {
+            wrappedValue = try loadValueFromKeychain()
+        } catch {
+            logError(error)
+        }
+    }
+
+    private func logError(_ error: Error) {
+        guard let keychainError = error as? KeychainError else {
+            return NSLog("An unexpected error occurred:\nKey: %@\nError: %@", key, error.localizedDescription)
+        }
+
+        NSLog(keychainError.description(for: key))
+    }
+
+    private func loadValueFromKeychain() throws -> Value {
+        var searchQuery = self.searchQuery
+        searchQuery[kSecReturnAttributes as String] = true
+        searchQuery[kSecReturnData as String] = true
+
+        let keychainResponse = keychain.fetch(searchQuery)
+
+        guard
+            keychainResponse.status != errSecItemNotFound,
+            keychainResponse.status == errSecSuccess,
+            let item = keychainResponse.queryResult as? [String: Any],
+            let data = item[kSecValueData as String] as? Data
+        else { throw KeychainError.itemNotFound }
+
+        do {
+            let value = try JSONDecoder().decode(Value.self, from: data)
+            return value
+        } catch {
+            throw KeychainError.decodingError(error: error)
+        }
+    }
+
+    private func storeValueInKeychain(_ value: Value?) throws {
+        guard let value = value else { return try deleteFromKeychain() }
+
+        let encodedData: Data
+        do {
+            encodedData = try JSONEncoder().encode(value)
+        } catch {
+            try deleteFromKeychain()
```

Why are you deleting the keychain entry in `storeValueInKeychain`? 

Maybe an app version has a problem with reading the value as a bug appeared with a data type mismatch, do you really want to delete the entry then or shall we keep it to have the possibility to fix this bug?

---

@JensK611 @mrylmz What do you think about that?

---

I totally agree to @SimonNumberTwo. In this case I would throw an error.  The developer who is using the lib should decide what is happening in an error case.